producer/writer/consumer_service_writer.go

```diff
@@ -280,23 +285,33 @@ func (w *consumerServiceWriterImpl) diffPlacementWithLock(newPlacement placement
 
 func (w *consumerServiceWriterImpl) Close() {
 	w.Lock()
+	defer w.Unlock()
+
 	if w.closed {
-		w.Unlock()
 		return
 	}
+	w.logger.Infof("closing consumer service writer %s", w.cs.String())
 	w.closed = true
-	w.Unlock()
-
 	close(w.doneCh)
+
 	// Blocks until all messages consuemd.
+	var shardWriterWG sync.WaitGroup
 	for _, sw := range w.shardWriters {
-		sw.Close()
+		sw := sw
+		shardWriterWG.Add(1)
+		go func() {
+			sw.Close()
+			shardWriterWG.Done()
+		}()
 	}
+	shardWriterWG.Wait()
```

Comment:
Make sure to check if there's a possibility for deadlock here if you are holding the consumer service lock while waiting for other goroutines to finish that may be waiting for the same lock.